### ISSUE TITLE
[Issue] azd pipeline config: "no oid claim"

### ISSUE THREAD

**Issue by pamelafox** [Azure/azure-dev#1758]
- [X] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**

azd version 0.7.0-beta.1 (commit 9ce71659f7688d0dc3dda8b84e5accedca58cf01)

**Describe the bug**

I am trying to use my personal MS account to set up a pipeline but keep getting an error. @weikanglim mentioned a friend having a similar issue.

```
(.venv) pamelafox-site % azd pipeline config
? Please enter a new environment name: pamelafox-swa
? Please select an Azure Subscription to use:  1. Azure subscription 1 (74feb6f3-f646-478d-a99b-37fafee75e29)
? Please select an Azure location to use: 38. (US) Central US (centralus)
Error: loading environment: initializing environment: fetching current user information: getting oid from token: no oid claim
```

With debug:

```
(.venv) ?1pamelafox-site % azd pipeline config --debug
2023/03/19 13:59:22 main.go:166: using cached latest version: 0.7.0-beta.1 (expires on: 2023-03-19T21:42:02Z)
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-config-list-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-config-get-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-config-set-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-config-unset-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-config-reset-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-env-set-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-set-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-select-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-env-new-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-new-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-list-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-env-refresh-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-refresh-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-env-get-values-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-env-get-values-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-infra-create-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-infra-create-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-infra-delete-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-infra-delete-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-pipeline-config-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-pipeline-config-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-telemetry-upload-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-template-list-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-template-show-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-auth-token-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-auth-token-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-version-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-version-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-show-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-show-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-restore-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-restore-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-login-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-login-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-logout-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-init-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-init-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-provision-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-provision-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-deploy-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-deploy-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-up-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-up-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-monitor-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-monitor-action'
2023/03/19 13:59:22 cobra_builder.go:170: registering flags for action 'azd-down-action'
2023/03/19 13:59:22 cobra_builder.go:190: registering resolver for action 'azd-down-action'
? Please enter a new environment name: pamelafox-swa
? Please select an Azure Subscription to use:  1. Azure subscription 1 (74feb6f3-f646-478d-a99b-37fafee75e29)
  |=      | Retrieving locations...2023/03/19 13:59:31 main.go:46: Retry: =====> Try=1
2023/03/19 13:59:31 main.go:46: Request: ==> OUTGOING REQUEST (Try=1)
   GET https://management.azure.com/subscriptions/74feb6f3-f646-478d-a99b-37fafee75e29/locations?api-version=2021-01-01
   Accept: application/json
   Authorization: REDACTED
   User-Agent: azsdk-go-armsubscriptions/v1.0.0 (go1.20; darwin),azdev/0.7.0-beta.1 (Go go1.20; darwin/amd64)

  |====   | Retrieving locations...2023/03/19 13:59:32 main.go:46: Response: ==> REQUEST/RESPONSE (Try=1/653.969333ms, OpTime=654.56375ms) -- RESPONSE RECEIVED
   GET https://management.azure.com/subscriptions/74feb6f3-f646-478d-a99b-37fafee75e29/locations?api-version=2021-01-01
   Accept: application/json
   Authorization: REDACTED
   User-Agent: azsdk-go-armsubscriptions/v1.0.0 (go1.20; darwin),azdev/0.7.0-beta.1 (Go go1.20; darwin/amd64)
   --------------------------------------------------------------------------------
   RESPONSE Status: 200 OK
   Cache-Control: no-cache
   Content-Type: application/json; charset=utf-8
   Date: Sun, 19 Mar 2023 20:59:31 GMT
   Expires: -1
   Pragma: no-cache
   Strict-Transport-Security: REDACTED
   Vary: REDACTED
   X-Content-Type-Options: REDACTED
   X-Ms-Correlation-Request-Id: REDACTED
   X-Ms-Ratelimit-Remaining-Subscription-Reads: REDACTED
   X-Ms-Request-Id: 95f5ec6a-6359-47a3-9fda-233514233311
   X-Ms-Routing-Request-Id: REDACTED

2023/03/19 13:59:32 main.go:46: Retry: response 200
2023/03/19 13:59:32 main.go:46: Retry: exit due to non-retriable status code
? Please select an Azure location to use: 38. (US) Central US (centralus)
2023/03/19 13:59:34 main.go:46: Retry: =====> Try=1
2023/03/19 13:59:34 main.go:46: Request: ==> OUTGOING REQUEST (Try=1)
   GET https://graph.microsoft.com/v1.0/me
   Authorization: REDACTED
   User-Agent: azsdk-go-graph/1.0.0 (go1.20; darwin),azdev/0.7.0-beta.1 (Go go1.20; darwin/amd64)

2023/03/19 13:59:34 main.go:46: Response: ==> REQUEST/RESPONSE (Try=1/183.701333ms, OpTime=183.775208ms) -- RESPONSE RECEIVED
   GET https://graph.microsoft.com/v1.0/me
   Authorization: REDACTED
   User-Agent: azsdk-go-graph/1.0.0 (go1.20; darwin),azdev/0.7.0-beta.1 (Go go1.20; darwin/amd64)
   --------------------------------------------------------------------------------
   RESPONSE Status: 403 Forbidden
   Cache-Control: no-cache
   Client-Request-Id: REDACTED
   Content-Type: application/json
   Date: Sun, 19 Mar 2023 20:59:33 GMT
   Request-Id: 139e40e5-50f3-4288-9d63-cfe9cd28e506
   Strict-Transport-Security: REDACTED
   Vary: REDACTED
   X-Ms-Ags-Diagnostic: REDACTED

2023/03/19 13:59:34 main.go:46: Retry: response 403
2023/03/19 13:59:34 main.go:46: Retry: exit due to non-retriable status code
Error: loading environment: initializing environment: fetching current user information: getting oid from token: no oid claim
```

**To Reproduce**

You'd likely need to switch from corp MS to personal MS. Not sure what other factors are at play.

**Environment**
Information on your environment: Mac M1 Ventura

**Comment by pamelafox** [comment]
No, I didn't find a workaround yet, I tried `azd login` a few times. I was hoping you might have a workaround if I posted it. ;-) Maybe I should blow out the top level azure directory?

**Comment by vhvb1989** [comment]
The root cause is that azd uses `scope=https%3a%2f%2fmanagement.azure.com%2f%2f.default+openid+offline_access+profile` which won't grant MSGraph access.
MSGraph is required for `azd pipeline config` because it creates a SP.

@pamelafox, here is a workaround you can try.  It is a little ugly, but please lmk if it works for you.

1. Run `azd login`
2. When the WebBroswer opens, do not select a credential or do login. Instead, copy and paste the entire `url` from the browser to some editor, like VSCode or NotePad.
3. Within the url, find this: `scope=https%3a%2f%2fmanagement.azure.com%2f%2f.default+openid+offline_access+profile` and replace it for this: `scope=https%3a%2f%2fgraph.microsoft.com%2f.default+offline_access+openid+profile`
4. After replacing, copy the url and paste it into the WebBrowser and hit enter. Continue the login process normally

This will give msgraph access to azd for running msgraph requests.

You might be wondering, why does it work with my Msft account??   And this is only because you probably run `az login scope=https://graph.microsoft.com/.default` in the past and used your msft account.  After that, `azd login` or `az login` will continue have the msgraph scope.

**Comment by vhvb1989** [comment]
Yes, `azd` require both scopes, but AFAIK, we can't request both scopes during login. And if we switch to graph, then deployment manager would fail.
But scopes are accumulative, so, once you login to one scope, and later to another, you keep access to both. For example, if you run `azd logout` and then `azd login` again to the same account, you don't need to re-apply the scope-workaround again... as the credential you select on the browser will just return the cached credentials with both scopes.

This is something we need to follow up with Azure Identity, and figure out if we can automatically request scopes after user-login without user interaction.

In the case of the `az cli`, then this happens, folks get an error and at the end of the error the user is asked to tun `az login --scope graph.microsoft.com/default` to fix this.  But I hope we can do it better on azd

**Comment by pamelafox** [comment]
Ahh okay, thanks for the explanation! Yeah, it'd be nice if there was a way to get more scopes. If this is going to be a common issue for folks when azd launches, maybe it could be documented on the troubleshooting page? (Not sure how much of an edge case my situation is.)

**Comment by vhvb1989** [comment]
I think @weikanglim is adding `--scope` support, which would offer the same experience than `az`
https://github.com/Azure/azure-dev/pull/1571

**Comment by vhvb1989** [comment]
related: https://github.com/Azure/azure-dev/issues/1516

**Comment by weikanglim** [comment]
@pamelafox Great that @vhvb1989 provided a workaround that helped!

@vhvb1989 are we okay with continuing this discussion on #1516?

**Comment by appliedcognetics** [comment]
I previously installed this repo with azd up https://github.com/Azure-Samples/azure-search-openai-demo
I did an azd down and now when I try to recreate the environment I get . I am using the VS Subscription Azure account , all of this worked fine the first time.
ERROR: loading environment: initializing environment: fetching current user information: getting oid from token: no oid claim

**Comment by weikanglim** [comment]
Will be fixed with #2185
